datastore/src/main/java/com/ds/datastore/BookStoreController.java

```diff
@@ -3,78 +3,229 @@
 import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
 import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;
 
+import java.io.BufferedReader;
+import java.io.DataOutputStream;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import javax.annotation.PostConstruct;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.hateoas.CollectionModel;
 import org.springframework.hateoas.EntityModel;
 import org.springframework.hateoas.IanaLinkRelations;
-import org.springframework.http.ResponseEntity;
 import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.*;
-
-import java.util.List;
-import java.util.stream.Collectors;
+import org.springframework.web.util.UriComponentsBuilder;
 
 @RestController
 public class BookStoreController {
 
-    private final BookStoreRepository repository;
     private final BookStoreModelAssembler assembler;
+    private final BookRepository bookRepository;
+    private final BookStoreRepository storeRepository;
+    private ServerMap map;
+    private Long id = null;
+
+    @Value("${application.baseUrl}")
+    private String url;
 
-    public BookStoreController(BookStoreRepository repository, BookStoreModelAssembler assembler) {
-        this.repository = repository;
+    public BookStoreController(BookStoreModelAssembler assembler, BookRepository bookRepository, BookStoreRepository storeRepository, ServerMap map) {
         this.assembler = assembler;
+        this.bookRepository = bookRepository;
+        this.storeRepository = storeRepository;
+        this.map = map;
     }
 
-    @PostMapping("/bookstores")
-    protected ResponseEntity<EntityModel<BookStore>> newBookStore(@RequestBody BookStore bookStore){
-        EntityModel<BookStore> entityModel = assembler.toModel(repository.save(bookStore));
+    @PostConstruct
+    private void restartChangedOrNew() throws Exception {
+        this.map.setMap(reclaimMap());
+        List<BookStore> bookStoreList = storeRepository.findAll();
+        if(!bookStoreList.isEmpty()) {
+            BookStore bookStore = bookStoreList.get(0);
+            this.id = bookStore.getServerId();
+            String serverAddress = this.map.get(this.id);
+            if(!serverAddress.equals(url + "/bookstores/" + this.id)) {
```

Comment:
if I am reading this correctly, would this mean that if the port changes, it will be assigned a new server id? That doesnt seem desirable.